uaparser/parser.go

```diff
@@ -146,119 +152,83 @@ type Parser struct {
-	cDefaultCacheSize      = 1024
+	EOsLookUpMode          LookupMode = 1 /* 00000001 */
+	EUserAgentLookUpMode   LookupMode = 2 /* 00000010 */
+	EDeviceLookUpMode      LookupMode = 4 /* 00000100 */
+	cMinMissesTreshold                = 100000
+	cDefaultMissesTreshold            = 500000
+	cDefaultMatchIdxNotOk             = 20
+	cDefaultSortOption                = false
+	cDefaultDebugMode                 = false
+	cDefaultCacheSize                 = 1024
 )
 
-func (parser *Parser) mustCompile() { // until we can use yaml.UnmarshalYAML with embedded pointer struct
-	for _, p := range parser.UA {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
-	}
-	for _, p := range parser.OS {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
-	}
-	for _, p := range parser.Device {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
+func New(options ...Option) (*Parser, error) {
+	parser := &Parser{
+		config: &parserConfig{
+			Mode:            EOsLookUpMode | EUserAgentLookUpMode | EDeviceLookUpMode,
+			UseSort:         cDefaultSortOption,
+			DebugMode:       cDefaultDebugMode,
+			CacheSize:       cDefaultCacheSize,
+			MissesThreshold: cMinMissesTreshold,
+			MatchIdxNotOk:   cDefaultMatchIdxNotOk,
+		},
+		mu: &sync.RWMutex{},
 	}
-}
 
-func defaultParserConfig() *parserConfig {
-	return &parserConfig{
-		Mode: EOsLookUpMode | EUserAgentLookUpMode | EDeviceLookUpMode,
-		UseSort: cDefaultSortOption,
-		DebugMode: cDefaultDebugMode,
-		CacheSize: cDefaultCacheSize,
-		MissesThreshold: cMinMissesTreshold,
-		MatchIdxNotOk:   cDefaultMatchIdxNotOk,
+	for _, o := range options {
+		o(parser)
 	}
-}
 
-func NewWithOptions(regexFile string, mode, treshold, topCnt int, useSort, debugMode bool, cacheSize int) (*Parser, error) {
-	data, err := os.ReadFile(regexFile)
-	if nil != err {
-		return nil, err
+	if parser.config.MatchIdxNotOk < 0 {
+		parser.config.MatchIdxNotOk = 0
 	}
 
-	cfg := &parserConfig{
-		Mode: mode,
-		UseSort: useSort,
-		DebugMode: debugMode,
-		MatchIdxNotOk: cDefaultMatchIdxNotOk,
-		MissesThreshold: cDefaultMissesTreshold,
-		CacheSize: cDefaultCacheSize,
+	if parser.config.MissesThreshold <= cMinMissesTreshold {
+		parser.config.MissesThreshold = cMinMissesTreshold
 	}
 
-	if topCnt >= 0 {
-		cfg.MatchIdxNotOk = topCnt
-	}
-	if treshold > cMinMissesTreshold {
-		cfg.MissesThreshold = uint64(treshold)
-	}
-	if cacheSize > 0 {
-		cfg.CacheSize = cacheSize
+	if parser.config.CacheSize < 0 {
+		parser.config.CacheSize = cDefaultCacheSize
 	}
 
-	parser, err := newFromBytes(data, cfg)
-	if err != nil {
-		return nil, err
+	if parser.cache == nil {
+		parser.cache = newCache(parser.config.CacheSize)
 	}
-	return parser, nil
-}
 
-func New(regexFile string) (*Parser, error) {
-	data, err := os.ReadFile(regexFile)
-	if nil != err {
-		return nil, err
+	if parser.RegexDefinitions == nil {
+		regexesDefinitions := defaultRegexesDefinitions()
+		parser.RegexDefinitions = &regexesDefinitions
```

Comment:
golang memory model question: is this doing a copy of the `defaultRegexesDefinition()` return value? we only need to copy in the case that `UseSort` is true, which is the rare case - as I'd expect most users of the library aren't using that option.  If we could only copy if that is in use that would be better.

Reply:
> is this doing a copy of the defaultRegexesDefinition() return value?

Yes. `defaultRegexesDefinition()` now returns a value (not a pointer) which results into a copy. Looking deeper, it's not a fully copy. In go, slices are always references. In that case, only the slice headers are copied, but all copies references the same data. 

Thats means, if `defaultRegexesDefinition()` is muted, it's muted everywhere. 

Hm. seems like we avoid the sync.Once.

Reply:
better, but it's still not quite right... unfortunately the MatchesCount fields from `uaParser`, `osParser`, and `deviceParser` are each mutated during the corresponding `Parse*` functions... meaning where we before had independent `MatchesCounts` per `Parser`, we now share those across `Parser`s when the same regex definitions are in use.

... those are only used to do anything in the UseSort case, so I think it'd be simpler instead of copy right before sort, to copy in `New` but only if `UseSort` is used, than to copy in `checkAndSort`.